presto-spi/src/main/java/com/facebook/presto/spi/block/StructBuilder.java

```diff
@@ -0,0 +1,214 @@
+    {
+        if (block.isNull(position)) {
+            builder.appendNull();
+        }
+        else {
+            block.writeBytesTo(position, 0, block.getLength(position), builder);
+            builder.closeEntry();
+        }
+        return this;
+    }
+
+    public StructBuilder addAll(Block block)
+    {
+        for (int i = 0; i < block.getPositionCount(); i++) {
+            add(block, i);
+        }
+
+        return this;
+    }
+
+    public StructBuilder add(Object value)
+    {
+        if (!typeIterator.hasNext()) {
+            throw new IllegalArgumentException("Cannot add element on a unknown type");
+        }
+        Type type = typeIterator.next();
+        if (value == null) {
+            builder.appendNull();
+            return this;
+        }
+        if (type.getJavaType() == boolean.class) {
+            type.writeBoolean(builder, (boolean) value);
+        }
+        else if (type.getJavaType() == long.class) {
+            if (value instanceof Integer) {
+                type.writeLong(builder, (int) value);
+            }
+            else if (value instanceof Long) {
+                type.writeLong(builder, (long) value);
+            }
+        }
+        else if (type.getJavaType() == double.class) {
+            if (value instanceof Float) {
+                type.writeDouble(builder, ((Float) value).doubleValue());
+            }
+            else if (value instanceof Double) {
+                type.writeDouble(builder, (double) value);
+            }
+        }
+        else if (type.getJavaType() == Slice.class) {
+            Slice slice;
+            if (value instanceof String) {
+                slice = Slices.utf8Slice((String) value);
+            }
+            else if (value instanceof Slice) {
+                slice = (Slice) value;
+            }
+            else if (value instanceof List) {
+                StructBuilder structBuilder = new StructBuilder(type.getTypeParameters());
+                ((List<?>) value).forEach(structBuilder::add);
+                slice = structBuilder.build();
+            }
+            else if (value instanceof Map) {
+                StructBuilder structBuilder = new StructBuilder(type.getTypeParameters());
+                ((Map<?, ?>) value).forEach((k, v) -> structBuilder.add(k).add(v));
+
+                slice = structBuilder.build();
+            }
+            else if (value instanceof byte[]) {
+                slice = Slices.wrappedBuffer((byte[]) value);
+            }
+            else {
+                throw new UnsupportedOperationException("Unsupported type " + value.getClass());
+            }
+            type.writeSlice(builder, slice);
+        }
+        else {
+            throw new UnsupportedOperationException("Unsupported type " + type);
+        }
+
+        return this;
+    }
+
+    private class CycleIterator implements Iterator<Type>
+    {
+        int index = 0;
+        final int size;
```

Comment:
This can be private. Also, swap the order so the mutable variable comes after the final one.